src/system/DurationTimer.cpp

```diff
@@ -0,0 +1,132 @@
+
+#include <stdint.h>
+#include <string>
+#include <system/DurationTimer.h>
+
+#if CHIP_DEVICE_USES_SYS_TIME
+#include <lib/support/logging/CHIPLogging.h>
+#include <system/SystemClock.h>
+#endif
+
+#if CHIP_DEVICE_USES_TIME_H
+#include <time.h>
+#endif
+
+using namespace std;
+using namespace std::literals;
+
+namespace chip {
+namespace timing {
+
+#define DATETIME_PATTERN ("%Y-%m-%dT%H:%M:%S")
+
+#define DATETIME_LEN (sizeof "1970-01-01T23:59:59")
+
+#define ISO8601_LEN (sizeof "1970-01-01T23:59:59.123456Z")
+
+#ifdef CHIP_DEVICE_USES_SYS_TIME
+// member functions
+void DurationTimer::start()
+{
+    chip::System::Clock::Timestamp tv1 = chip::System::SystemClock().GetMonotonicTimestamp();
+
+    chip::System::Clock::ToTimeval(tv1, t1);
+
+    ChipLogDetail(DeviceLayer, "Timer: %s start %s ", label.c_str(), toTimeStr(&t1).c_str());
+}
+
+void DurationTimer::stop()
+{
+
+    chip::System::Clock::Timestamp tv2 = chip::System::SystemClock().GetMonotonicTimestamp();
+
+    chip::System::Clock::ToTimeval(tv2, t2);
+
+    ChipLogDetail(DeviceLayer, "Timer: %s stop %s ", label.c_str(), toTimeStr(&t2).c_str());
+
+    duration();
+}
+
+double DurationTimer::duration()
+{
+    double dur = ((t2.tv_usec - t1.tv_usec) * 1e-6);
+
+    string timestr = toTimeStr(&t2);
+
+    ChipLogDetail(DeviceLayer, "Timer: %s TIME_SPENT (sec) %f , %s ", label.c_str(), dur, timestr.c_str());
+
+    return dur;
+}
+
+// utility method
+string DurationTimer::toTimeStr(timeval * time)
+{
+    char * buff = new char[DATETIME_LEN];
+
+    struct tm * tm_info = gmtime(&(time->tv_sec));
+    strftime(buff, DATETIME_LEN, DATETIME_PATTERN, tm_info);
+    char * str = new char[ISO8601_LEN];
+    snprintf(str, ISO8601_LEN, "time: %s.%05ld", buff, time->tv_usec);
+
+    return str;
+}
+#endif
+
+#ifdef CHIP_DEVICE_USES_TIME_H
```

Comment:
Why wouldn't we just always implement this on top of `GetMonotonicTimestamp()`?   If that happens to be CLOCK_REALTIME, fine, but for purposes of this class it should not matter, right?